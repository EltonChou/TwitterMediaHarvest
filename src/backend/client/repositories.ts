import { Sha256 } from '@aws-crypto/sha256-browser'
import type { CognitoIdentityCredentials } from '@aws-sdk/credential-provider-cognito-identity'
import { FetchHttpHandler, streamCollector } from '@aws-sdk/fetch-http-handler'
import { HttpRequest } from '@aws-sdk/protocol-http'
import { SignatureV4 } from '@aws-sdk/signature-v4'
import { CreateClientFailed, UpdateStatsFailed } from '@backend/errors'
import type { IStorageProxy } from '@libs/proxy'
import type { ClientInfo, V4Statistics } from '@schema'
import jws from 'jws'
import type { Storage } from 'webextension-polyfill'
import Browser from 'webextension-polyfill'
import { ClientInfoVO } from './valueObjects'

type UpdateInfo = Pick<ClientInfo, 'csrfToken' | 'syncedAt'>

const defaultInfo: Readonly<ClientInfo> = {
  uuid: 'uuid',
  csrfToken: 'token',
  syncedAt: 0,
  uninstallCode: 'code',
}

const isEmptyInfo = (record: Record<string, unknown>) =>
  record['uuid'] === defaultInfo.uuid ||
  record['csrfToken'] === defaultInfo.csrfToken ||
  record['syncedAt'] === defaultInfo.syncedAt ||
  record['uninstall'] === defaultInfo.uninstallCode

type ProviderOptions = {
  credentialProvider: Provider<CognitoIdentityCredentials>
  statsProvider: Provider<V4Statistics>
}

export interface IClientInfoRepository {
  getInfo(options?: ProviderOptions): Promise<ClientInfoVO>
  updateStats(options?: ProviderOptions): Promise<void>
  resetInfo(): Promise<void>
}

export class InfoSyncLock implements IProcessLock {
  private lockCriteria = 'InfoUpdateLock'
  private maxLockTime = 10 * 60 * 1000

  constructor(readonly storageArea: Storage.StorageArea) {}

  async isLocked(): Promise<boolean> {
    const record = await this.storageArea.get(this.lockCriteria)
    return Object.keys(record).includes(this.lockCriteria) && record[this.lockCriteria] - Date.now() >= this.maxLockTime
  }

  async release(): Promise<void> {
    await this.storageArea.remove(this.lockCriteria)
  }

  async acquire(): Promise<void> {
    await this.storageArea.set({ [this.lockCriteria]: Date.now() })
  }
}

export class ClientInfoRepository implements IClientInfoRepository {
  constructor(readonly storageArea: IStorageProxy<ClientInfo>, private defaultOptions: ProviderOptions) {}

  private async createClient(options: ProviderOptions): Promise<ClientTokenResponse> {
    const initStats = await options.statsProvider()
    const credential = await options.credentialProvider()

    const handler = new ClientApiHandler()
    const response = await handler.send(HttpMethod.Post, '/clients', {}, initStats, credential)

    if (response.statusCode === 201) {
      const body = JSON.parse(response.body)
      return body
    }

    throw new CreateClientFailed(
      response.statusCode,
      response.body?.message || JSON.stringify(response.body),
      response.headers
    )
  }

  private async getNewInfo(options?: Partial<ProviderOptions>): Promise<ClientInfoVO> {
    const clientTokenResponse: ClientTokenResponse = await this.createClient({ ...this.defaultOptions, ...options })
    const { payload } = jws.decode(clientTokenResponse.token, { json: true })
    const info = new ClientInfoVO({
      uuid: payload['uuid'],
      csrfToken: clientTokenResponse.token,
      syncedAt: Date.now(),
      uninstallCode: clientTokenResponse?.uninstallCode || payload['uninstallCode'],
    })
    await this.storageArea.setItem(info.props)
    await Browser.runtime.setUninstallURL(info.uninstallUrl)
    return info
  }

  async getInfo(options?: Partial<ProviderOptions>): Promise<ClientInfoVO> {
    const record = await this.storageArea.getItemByDefaults(defaultInfo)
    return isEmptyInfo(record) ? this.getNewInfo(options) : new ClientInfoVO(record)
  }

  async updateStats(options?: Partial<ProviderOptions>): Promise<void> {
    options = { ...this.defaultOptions, ...options }

    const { uuid, csrfToken } = await this.getInfo(options)
    const stats = await options.statsProvider()
    const credential = await options.credentialProvider()

    const handler = new ClientApiHandler()
    const response = await handler.send(
      HttpMethod.Put,
      '/clients/' + uuid + '/stats',
      { 'x-csrf-token': csrfToken },
      stats,
      credential
    )

    if (response.statusCode !== 200)
      throw new UpdateStatsFailed(
        response.statusCode,
        response.body?.message || JSON.stringify(response.body),
        response.headers
      )

    const body = JSON.parse(response.body)
    const clientInfo: UpdateInfo = {
      csrfToken: body.token,
      syncedAt: Date.now(),
    }
    await this.storageArea.setItem(clientInfo)
  }

  async resetInfo(): Promise<void> {
    await this.storageArea.setItem(defaultInfo)
  }
}

enum HttpMethod {
  Post = 'POST',
  Get = 'GET',
  Put = 'PUT',
  Delete = 'DELETE',
}

class ClientApiHandler {
  async send(
    method: HttpMethod,
    path: string,
    headers: Record<string, string>,
    body: unknown,
    credentials: CognitoIdentityCredentials
  ) {
    const request = new HttpRequest({
      protocol: 'https',
      body: JSON.stringify(body),
      headers: {
        'Content-Type': 'application/json',
        'X-Api-Key': process.env.API_KEY,
        'X-MediaHarvest-Version': process.env.VERSION_NAME,
        host: process.env.API_HOSTNAME,
        ...headers,
      },
      method: method,
      hostname: process.env.API_HOSTNAME,
      path: (process.env.API_ROOT_PATH || '') + path,
    })

    const signer = new SignatureV4({
      region: 'ap-northeast-1',
      service: 'execute-api',
      credentials: credentials,
      sha256: Sha256,
    })

    const signedRequest = await signer.sign(request)

    const client = new FetchHttpHandler()
    const { response } = await client.handle(signedRequest as HttpRequest)

    response.body = new TextDecoder().decode(await streamCollector(response.body))
    return response
  }
}
