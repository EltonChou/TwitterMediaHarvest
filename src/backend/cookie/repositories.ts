import browser from 'webextension-polyfill'

export interface ITwitterTokenRepository {
  getCsrfToken(): Promise<string | null>
  getGuestToken(): Promise<string | null>
}

/**
 * @deprecated
 */
export class TwitterTokenRepository implements ITwitterTokenRepository {
  async getCsrfToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://twitter.com',
      name: 'ct0',
    })

    return cookie ? cookie.value : null
  }

  async getGuestToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://twitter.com',
      name: 'gt',
    })
    return cookie ? cookie.value : null
  }
}

export class XTokenRepository implements ITwitterTokenRepository {
  async getCsrfToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://x.com',
      name: 'ct0',
    })

    return cookie ? cookie.value : null
  }

  async getGuestToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://x.com',
      name: 'gt',
    })
    return cookie ? cookie.value : null
  }
}
