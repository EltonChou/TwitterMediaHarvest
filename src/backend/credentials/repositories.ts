import type { CognitoIdentityCredentials, Storage } from '@aws-sdk/credential-provider-cognito-identity'
import { fromCognitoIdentityPool } from '@aws-sdk/credential-providers'
import ValueObject from '@backend/valueObject'
import type { AWSCredentials } from '@schema'
import type { Storage as BrowserStorage } from 'webextension-polyfill'
import Browser from 'webextension-polyfill'

class CredentialVO extends ValueObject<AWSCredentials> {
  readonly expireTimeRedundancy: number = 3 * 60 * 1000

  constructor(props: AWSCredentials) {
    super(props)
  }

  get credential(): CognitoIdentityCredentials {
    return { ...this.props, expiration: new Date(this.props.expiration) }
  }

  get isExpired(): boolean {
    return this.props?.expiration ? this.props.expiration < Date.now() - this.expireTimeRedundancy : true
  }

  static fromCognitoIdentityCredentials(credential: CognitoIdentityCredentials): CredentialVO {
    return new CredentialVO({
      ...credential,
      // ? Does it means the credential wouldn't expired if the expiration was undefined?
      expiration: credential?.expiration ? credential.expiration.getTime() : 0,
    })
  }
}

export interface ICredentialRepository {
  getCredential(): Promise<CognitoIdentityCredentials>
}

class BrowserSyncStorageProxy implements Storage {
  async getItem(key: string): Promise<string | null> {
    const record = await Browser.storage.sync.get(key)
    return Object.keys(record).length === 0 ? null : record[key]
  }

  async removeItem(key: string): Promise<void> {
    return await Browser.storage.sync.remove(key)
  }

  async setItem(key: string, data: string): Promise<void> {
    return await Browser.storage.sync.set({ [key]: data })
  }
}

export class CredentialRepository implements ICredentialRepository {
  readonly credentialCriteria = 'awsCredential'

  constructor(readonly storageArea: BrowserStorage.StorageArea) {}

  private async fetchCredential(): Promise<CognitoIdentityCredentials> {
    // ! Data in storage is `AWSCredential` not `CognitoIdentityCredentials`
    // TODO: Need to define the schema of storage area.
    const credential = await fromCognitoIdentityPool({
      identityPoolId: process.env.IDENTITY_POOL_ID,
      cache: new BrowserSyncStorageProxy(),
      clientConfig: {
        region: process.env.IDENTITY_POOL_REGION,
      },
    })()

    const credentialVO = CredentialVO.fromCognitoIdentityCredentials(credential)
    await this.storageArea.set({ [this.credentialCriteria]: credentialVO.props })
    return credential
  }

  async getCredential(): Promise<CognitoIdentityCredentials> {
    const credentialRecord = await this.storageArea.get(this.credentialCriteria)
    if (Object.keys(credentialRecord).length === 0) return await this.fetchCredential()

    const credentialsVO = new CredentialVO(credentialRecord[this.credentialCriteria])
    return credentialsVO.isExpired ? await this.fetchCredential() : credentialsVO.credential
  }
}
