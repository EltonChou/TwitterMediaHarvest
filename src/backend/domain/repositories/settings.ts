export interface ISettingsRepository<Settings> {
  get(): Promise<Settings>
  save(settings: Partial<Settings>): Promise<void>
  reset(): Promise<void>
  getDefault(): Settings
}
