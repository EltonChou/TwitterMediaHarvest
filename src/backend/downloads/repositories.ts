import { DownloadHistoryEntity, DownloadRecord } from './models'
import type { DownloadDBSchema, DownloadHistoryItem, HashtagItem } from '@schema'
import { type IDBPDatabase } from 'idb'
import type { Storage } from 'webextension-polyfill'

export interface IDownloadRecordsRepository {
  getById(downloadItemId: number): Promise<DownloadRecord | null>
  save(downloadRecord: DownloadRecord): Promise<void>
  removeById(downloadItemId: number): Promise<void>
  getAll(): Promise<DownloadRecord[]>
}

abstract class IndexedDBRepository {
  constructor(readonly clientProvider: Provider<IDBPDatabase<DownloadDBSchema>>) {}
}

export class IndexedDBDownloadRecordsRepository
  extends IndexedDBRepository
  implements IDownloadRecordsRepository
{
  async save(downloadRecord: DownloadRecord): Promise<void> {
    const client = await this.clientProvider()
    await client.put('record', downloadRecord.toJson())
  }

  async getById(downloadItemId: number): Promise<DownloadRecord | null> {
    const client = await this.clientProvider()
    const record = await client.get('record', downloadItemId)
    return record ? DownloadRecord.fromJson(record) : null
  }

  async removeById(downloadItemId: number): Promise<void> {
    const client = await this.clientProvider()
    await client.delete('record', downloadItemId)
  }

  async getAll(): Promise<DownloadRecord[]> {
    const client = await this.clientProvider()
    const records = await client.getAll('record')
    return records.map(DownloadRecord.fromJson)
  }
}

type DownloadRecordId = `dl_${number}`
const storageAreaDownloadIdPattern = /dl_(?:\d+)/

export class StorageAreaDownloadRecordsRepository implements IDownloadRecordsRepository {
  constructor(readonly storageArea: Storage.StorageArea) {}

  async save(downloadRecord: DownloadRecord): Promise<void> {
    await this.storageArea.set({
      [createId(downloadRecord.id)]: downloadRecord.toJson(),
    })
  }

  async getById(downloadItemId: number): Promise<DownloadRecord | null> {
    const recordId = createId(downloadItemId)
    const volume = await this.storageArea.get(recordId)
    if (recordId in volume) {
      return DownloadRecord.fromJson(volume[recordId])
    }
    return null
  }

  async removeById(downloadItemId: number): Promise<void> {
    const recordId = createId(downloadItemId)
    await this.storageArea.remove(recordId)
  }

  async getAll(): Promise<DownloadRecord[]> {
    const volumes = await this.storageArea.get(null)

    return Object.entries(volumes)
      .filter(([k, v]) => Boolean(k.match(storageAreaDownloadIdPattern)))
      .map(([k, v]) => DownloadRecord.fromJson(v))
  }
}

function createId(downloadItemId: number): DownloadRecordId {
  return `dl_${downloadItemId}`
}

export type DownloadItemPredicate = (item: DownloadHistoryItem) => boolean

export interface IDownloadHistoryRepository {
  count(): Promise<number>
  save(item: DownloadHistoryEntity): Promise<void>
  clear(): Promise<void>
  tweetHasDownloaded(tweetId: string): Promise<boolean>
  getAll(): Promise<DownloadHistoryEntity[]>
  getLatest(limit?: number): Promise<DownloadHistoryEntity[]>
  getByTweetId(tweetId: string): Promise<DownloadHistoryEntity | undefined>
  searchByUserName(userName: string, limit?: number): Promise<DownloadHistoryEntity[]>
  search(
    ...predicates: DownloadItemPredicate[]
  ): (limit?: number, skip?: number) => Promise<[DownloadHistoryEntity[], number]>
}

export class IndexedDBDownloadHistoryRepository
  extends IndexedDBRepository
  implements IDownloadHistoryRepository
{
  private openCursor(
    range?: IDBKeyRange,
    direction?: IDBCursorDirection,
    limit: number | 'inf' = 50,
    offset = 0
  ) {
    const transaction = this.transaction()
    const shouldContinue = (count: number) => (limit === 'inf' ? true : count < limit)

    return async function* (indexKey: keyof DownloadDBSchema['history']['indexes']) {
      let cursor = await (await transaction).store
        .index(indexKey)
        .openCursor(range, direction)

      if (offset) cursor = await cursor.advance(offset)
      let count = 0
      while (cursor && shouldContinue(count)) {
        yield cursor.value
        cursor = await cursor.continue()
        count += 1
      }
    }
  }

  async transaction(mode: IDBTransactionMode = 'readonly') {
    const client = await this.clientProvider()
    const tx = client.transaction('history', mode)
    return tx
  }

  async tweetHasDownloaded(tweetId: string): Promise<boolean> {
    const client = await this.clientProvider()
    const key = await client.getKey('history', tweetId)
    return Boolean(key)
  }

  async clear(): Promise<void> {
    const client = await this.clientProvider()
    await client.clear('history')
  }

  async save(item: DownloadHistoryEntity): Promise<void> {
    const client = await this.clientProvider()
    await client.put('history', item.toDownloadHistoryItem())
  }

  async count(): Promise<number> {
    const client = await this.clientProvider()
    return client.count('history')
  }

  async getAll(): Promise<DownloadHistoryEntity[]> {
    const client = await this.clientProvider()
    const items = await client.getAll('history')
    return items.map(DownloadHistoryEntity.build)
  }

  async getLatest(limit = 10, offset = 0): Promise<DownloadHistoryEntity[]> {
    const collection = this.openCursor(null, 'prev', limit, offset)('byDownloadTime')
    return (await asyncGeneratorToArray(collection)).map(DownloadHistoryEntity.build)
  }

  async getByTweetId(tweetId: string): Promise<DownloadHistoryEntity | undefined> {
    const client = await this.clientProvider()
    const item = await client.get('history', tweetId)
    return item ? DownloadHistoryEntity.build(item) : undefined
  }

  async searchByUserName(userName: string, limit = 10): Promise<DownloadHistoryEntity[]> {
    const client = await this.clientProvider()
    const items = await client.getAll('history')
    const lowerCaseName = userName.toLowerCase()
    return items
      .filter(
        item =>
          item.displayName.toLowerCase().includes(lowerCaseName) ||
          item.screenName.toLowerCase().includes(lowerCaseName)
      )
      .map(DownloadHistoryEntity.build)
      .splice(0, limit)
  }

  search(...predicates: DownloadItemPredicate[]) {
    return async (
      limit?: number,
      skip?: number
    ): Promise<[DownloadHistoryEntity[], number]> => {
      const collection = this.openCursor(null, 'prev', 'inf')('byDownloadTime')
      const result: DownloadHistoryEntity[] = []

      let matchedCount = 0
      const increaseCount = () => (matchedCount += 1)
      for await (const v of collection) {
        if (Array.from(predicates).every(predicate => predicate(v))) {
          if (skip) {
            skip -= 1
            increaseCount()
            continue
          }
          if (result.length < limit) result.push(DownloadHistoryEntity.build(v))
          increaseCount()
        }
      }

      return [result, matchedCount]
    }
  }
}

const asyncGeneratorToArray = async <T>(
  generator: AsyncGenerator<T, void, unknown>
): Promise<T[]> => {
  const result: T[] = []
  for await (const v of generator) result.push(v)
  return result
}

export interface ITweetHashtagRepository {
  getAllTags(): Promise<string[]>
  getAll(): Promise<HashtagItem[]>
  getTweetsByTag(tag: string): Promise<string[]>
  getTweetsByTags(...tags: string[]): Promise<string[]>
  addTweet(tweetId: string): (...tag: string[]) => Promise<void>
  clear(): Promise<void>
}

export class IndexedDBHashtagRepository
  extends IndexedDBRepository
  implements ITweetHashtagRepository
{
  async getAllTags(): Promise<string[]> {
    const client = await this.clientProvider()
    return client.getAllKeys('hashtag')
  }

  async getAll(): Promise<HashtagItem[]> {
    const client = await this.clientProvider()
    return client.getAll('hashtag')
  }

  async getTweetsByTag(tag: string): Promise<string[]> {
    const client = await this.clientProvider()
    const item = await client.get('hashtag', tag)
    return item ? Array.from(item.tweetIds) : []
  }

  async getTweetsByTags(...tags: string[]): Promise<string[]> {
    const ids = new Set<string>()
    const client = await this.clientProvider()
    for (const tag of tags) {
      const item = await client.get('hashtag', tag)
      item && item.tweetIds.forEach(id => ids.add(id))
    }
    return Array.from(ids)
  }

  addTweet(tweetId: string): (...tags: string[]) => Promise<void> {
    return async (...tags: string[]) => {
      if (tags.length === 0) return
      const client = await this.clientProvider()
      for (const tag of tags) {
        const item = await client.get('hashtag', tag)
        const ids = item?.tweetIds || new Set()
        await client.put('hashtag', {
          name: tag,
          tweetIds: ids.add(tweetId),
        })
      }
    }
  }

  async clear() {
    const client = await this.clientProvider()
    await client.clear('hashtag')
  }
}
