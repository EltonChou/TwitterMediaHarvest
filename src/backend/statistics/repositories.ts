import {
  BrowserStorageFetcher,
  BrowserStorageSetter,
  storageFetcher,
  storageSetter
} from '../../libs/chromeApi'

export enum StatisticsKey {
  SuccessDownloadCount = 'successDownloadCount',
  FailedDownloadCount = 'failedDownloadCount',
  ErrorCount = 'errorCount',
}

type DownloadStatistic = {
  [StatisticsKey.SuccessDownloadCount]?: number
  [StatisticsKey.FailedDownloadCount]?: number
  [StatisticsKey.ErrorCount]?: number
}

const defaultStatistic: DownloadStatistic = {
  [StatisticsKey.ErrorCount]: 0,
  [StatisticsKey.FailedDownloadCount]: 0,
  [StatisticsKey.SuccessDownloadCount]: 0,
}

export interface IStatisticsRepository {
  getStatisticsCount(key: StatisticsKey): Promise<number>
  addStatisticsCount(key: StatisticsKey): Promise<void>
}

export default class StatisticsRepository implements IStatisticsRepository {
  private fetchStorage: BrowserStorageFetcher
  private setStorage: BrowserStorageSetter

  constructor(storageArea: chrome.storage.StorageArea) {
    this.fetchStorage = storageFetcher(storageArea)
    this.setStorage = storageSetter(storageArea)
  }

  async getStatisticsCount(
    key: StatisticsKey
  ): Promise<number> {
    const downloadStatistic: DownloadStatistic = {}
    downloadStatistic[key] = 0

    const count = await this.fetchStorage(downloadStatistic)
    return count[key]
  }

  async addStatisticsCount(key: StatisticsKey): Promise<void> {
    const count = await this.getStatisticsCount(key)
    const downloadCount: DownloadStatistic = {}
    downloadCount[key] = count + 1

    await this.setStorage(downloadCount)
  }

  async setDefaultStatistics(): Promise<void> {
    await this.setStorage(defaultStatistic)
  }
}