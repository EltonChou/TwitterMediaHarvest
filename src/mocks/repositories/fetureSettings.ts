import type { ISettingsRepository } from '#domain/repositories/settings'
import type { FeatureSettings } from '#schema'

export class MockFeatureSettingsRepository
  implements ISettingsRepository<FeatureSettings>
{
  protected settings: FeatureSettings
  constructor() {
    this.settings = {
      autoRevealNsfw: false,
      includeVideoThumbnail: false,
      keyboardShortcut: true,
    }
  }
  async get(): Promise<FeatureSettings> {
    return this.settings
  }
  async save(settings: Partial<FeatureSettings>): Promise<void> {
    this.settings = { ...this.settings, ...settings }
  }
  async reset(): Promise<void> {
    this.settings = this.getDefault()
  }
  getDefault(): FeatureSettings {
    return {
      autoRevealNsfw: false,
      includeVideoThumbnail: false,
      keyboardShortcut: true,
    }
  }
}
