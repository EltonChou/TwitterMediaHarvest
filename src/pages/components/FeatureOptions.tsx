import type { ISettingsRepository } from '#domain/repositories/settings'
import { getText as i18n } from '#libs/i18n'
import useFeatureSettings from '#pages/hooks/useFeatureSettings'
import type { FeatureSettings } from '#schema'
import DownloadKey from '../../contentScript/KeyboardMonitor/DownloadKey'
import { RichFeatureSwitch } from './controls/featureControls'
import { Kbd, Text, VStack } from '@chakra-ui/react'
import React from 'react'

const getKey = (downloadKey: DownloadKey): string => downloadKey.slice(-1)

const KeyboardShortcutDesc = () => {
  return (
    <Text as={'span'}>
      {i18n('Use keyboard shortcut to trigger download.', 'options:features')}
      <br />
      <Text as={'span'}>
        Twitter: <Kbd>{getKey(DownloadKey.Twitter)}</Kbd>
      </Text>
      <br />
      <Text as={'span'}>
        TweetDeck: <Kbd>{getKey(DownloadKey.BetaTweetDeck)}</Kbd>
      </Text>
    </Text>
  )
}

type FeatureOptionsProps = {
  featureSettingsRepo: ISettingsRepository<FeatureSettings>
}

const FeatureOptions = ({ featureSettingsRepo }: FeatureOptionsProps) => {
  const [featureSettings, toggler] = useFeatureSettings(featureSettingsRepo)

  return (
    <VStack>
      <RichFeatureSwitch
        name={i18n('Auto-reveal sensitive content', 'options:features')}
        desc={i18n(
          'When the tweet was flagged as sensitive content, this feature can show the blured content automatically.',
          'options:features'
        )}
        isOn={featureSettings.autoRevealNsfw}
        handleClick={toggler.nsfw}
        testId="revealNsfw-feature-switch"
      />
      <RichFeatureSwitch
        name={i18n('Keyboard shortcut', 'options:features')}
        desc={<KeyboardShortcutDesc />}
        isOn={featureSettings.keyboardShortcut}
        handleClick={toggler.keyboardShortcut}
        testId="keyboardShortcut-feature-switch"
      />
      <RichFeatureSwitch
        name={i18n('Download video thumbnail', 'options:features')}
        desc={i18n(
          'Download the thumbnail when the media is video.',
          'options:features'
        )}
        isOn={featureSettings.includeVideoThumbnail}
        handleClick={toggler.thumbnail}
        testId="videoThumbnail-feature-switch"
      />
    </VStack>
  )
}

export default FeatureOptions
