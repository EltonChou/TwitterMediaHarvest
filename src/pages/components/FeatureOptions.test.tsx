/**
 * @jest-environment jsdom
 */
import { MockFeatureSettingsRepository } from '#mocks/repositories/fetureSettings'
import FeatureOptions from './FeatureOptions'
import { queryAllByTestId, render } from '@testing-library/react'
import 'core-js/stable/structured-clone'
import React from 'react'

describe('unit test for FeatureOptions component', () => {
  const featureSettingsRepo = new MockFeatureSettingsRepository()

  afterAll(() => {
    jest.resetAllMocks()
    jest.restoreAllMocks()
  })

  it('can render properly', async () => {
    const { container, unmount } = render(
      <FeatureOptions featureSettingsRepo={featureSettingsRepo} />
    )

    expect(
      queryAllByTestId(container, new RegExp(/.*-feature-switch$/))
    ).toHaveLength(3)
    expect(container).toMatchSnapshot()

    unmount()
  })
})
