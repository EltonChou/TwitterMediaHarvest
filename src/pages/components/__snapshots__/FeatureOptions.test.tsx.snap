// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for FeatureOptions component can render properly 1`] = `
<div>
  <div
    class="chakra-stack css-tl3ftk"
  >
    <div
      class="chakra-form-control css-0"
      data-testid="revealNsfw-feature-switch"
      label="Translated<[options:features](Auto-reveal sensitive content)>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r0:"
        id="field-:r1:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;[options:features](Auto-reveal sensitive content)&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;[options:features](When the tweet was flagged as sensitive content, this feature can show the blured content automatically.)&gt;
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-switch__input"
                id=":r0:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
              >
                <span
                  class="chakra-switch__thumb css-0"
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
    <div
      class="chakra-form-control css-0"
      data-testid="keyboardShortcut-feature-switch"
      label="Translated<[options:features](Keyboard shortcut)>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r2:"
        id="field-:r3:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;[options:features](Keyboard shortcut)&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              <span
                class="chakra-text css-0"
              >
                Translated&lt;[options:features](Use keyboard shortcut to trigger download.)&gt;
                <br />
                <span
                  class="chakra-text css-0"
                >
                  Twitter: 
                  <kbd
                    class="chakra-kbd css-1u61e1q"
                  >
                    D
                  </kbd>
                </span>
                <br />
                <span
                  class="chakra-text css-0"
                >
                  TweetDeck: 
                  <kbd
                    class="chakra-kbd css-1u61e1q"
                  >
                    P
                  </kbd>
                </span>
              </span>
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-checked=""
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                checked=""
                class="chakra-switch__input"
                id=":r2:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
                data-checked=""
              >
                <span
                  class="chakra-switch__thumb css-0"
                  data-checked=""
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
    <div
      class="chakra-form-control css-0"
      data-testid="videoThumbnail-feature-switch"
      label="Translated<[options:features](Download video thumbnail)>"
      role="group"
    >
      <label
        class="chakra-form__label css-28a6i0"
        data-testid="feature-switch-label"
        for=":r4:"
        id="field-:r5:-label"
      >
        <div
          class="chakra-stack css-mz2o5n"
          style="transition: background 300ms;"
        >
          <div
            class="chakra-stack css-1bef4uc"
          >
            <p
              class="chakra-text css-1tbqbam"
            >
              Translated&lt;[options:features](Download video thumbnail)&gt;
            </p>
            <p
              class="chakra-text css-qljqz"
            >
              Translated&lt;[options:features](Download the thumbnail when the media is video.)&gt;
            </p>
             
          </div>
          <div
            class="css-fzqoy4"
          >
            <label
              class="chakra-switch css-16pgy8f"
              data-testid="feature-switch"
            >
              <input
                aria-disabled="false"
                aria-invalid="false"
                class="chakra-switch__input"
                id=":r4:"
                style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                type="checkbox"
                value=""
              />
              <span
                aria-hidden="true"
                class="chakra-switch__track css-14qxnv8"
              >
                <span
                  class="chakra-switch__thumb css-0"
                />
              </span>
            </label>
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;
